### PR TITLE
Ethereum bridge 0.1 + PoI Integration

### LINKED ISSUE #712: Initialize bridges subcrate
TBFE

### LINKED ISSUE #713: Implement Ethereum JSONRPC client in bridge
TBFE

### LINKED ISSUE #714: Subscribe bridge to new Witnet blocks
- Read unsolved data requests from WBI that are not scheduled for the future, check eligibility for each and post them to Witnet
- Iterate over data requests from the block received from Witnet, then claim each with PoI
- Iterate over tallies from the block received from Witnet, check eligibility for each and then post them to WBI with PoI

### LINKED ISSUE #728: [bridge] report blocks to relay

### LINKED ISSUE #766: Bridge PoI Integration
Modify bridge to handle new PoI ethereum contracts in  witnet/witnet-ethereum-bridge#13

### PR DESCRIPTION
**Pull request by lrubiorod**
Until PR https://github.com/witnet/witnet-ethereum-bridge/pull/13

This PR adds a new crate "witnet-ethereum-bridge" with some basic capabilities:

Close #712 
Close #713 
Close #714 
Close #728 
Close #766 

* Subscribe to and parse witnet blocks
* Subscribe to ethereum events

Error handling is not implemented. There is a configuration file which can be used to configure server addresses, contract address and ethereum account address.

The contract abi is stored as a separate file.

Sample run:

```
$ cargo run -p witnet-ethereum-bridge
[2019-06-26T09:51:15Z DEBUG witnet_ethereum_bridge] Loading config from `witnet_ethereum_bridge.toml`
[2019-06-26T09:51:15Z DEBUG witnet_ethereum_bridge] Web3 accounts: [0xfe1b2865b74200dceabf9ddee04fdc32c2c8e3fb, 0x62640139d722f483fd735a07fbff2d299b44c415, 0xd513883712d0cb037b12acd086e8c743f062a933, 0x12cb94056c05f79efecb2478f3d20e0ae4824259, 0x16e36bba89e48fcf16d3359d6500f7186e119724, 0xdb06bab06f187ab3ddfb18763628a3b096cce125, 0xdfd0146b967c1ef8f7477a92701b8874ea021ad3, 0x47b219b18c82df9013f38cb1e4989da8176da25e, 0xaab2867138bd70991e3bb8518a2947ed7ff0daa7, 0xc36681e64e959f16171748c3c5ceb1b5780dffd5]
[2019-06-26T09:51:15Z INFO  witnet_ethereum_bridge] Subscribing to contract 0x1aee6bad4e2b3ea2dadf6753da3d77da79290c9b topic 0xcf1533ed093ea4ecd1abdb7c4aeec88ad87b26afef4c722a681c8650227b90d5
[2019-06-26T09:51:15Z INFO  witnet_ethereum_bridge] Subscribed to witnet newBlocks with subscription id "3"
[2019-06-26T09:51:15Z DEBUG witnet_ethereum_bridge] Created filter: BaseFilter { id: "0x9", transport: Http { id: 2, url: http://127.0.0.1:8545/, basic_auth: None, write_sender: UnboundedSender(Sender { inner: Inner { buffer: None, state: 9223372036854775808, message_queue: Queue { head: 0x55dfb397e010, tail: UnsafeCell }, parked_queue: Queue { head: 0x7f0118030470, tail: UnsafeCell }, num_senders: 1, recv_task: Mutex { data: ReceiverTask { unparked: false, task: Some(Task) } } }, sender_task: Mutex { data: SenderTask { task: None, is_parked: false } }, maybe_parked: false }) }, item: PhantomData }
[2019-06-26T09:51:41Z DEBUG witnet_ethereum_bridge] Got ethereum event: Log { address: 0x1aee6bad4e2b3ea2dadf6753da3d77da79290c9b, topics: [0xcf1533ed093ea4ecd1abdb7c4aeec88ad87b26afef4c722a681c8650227b90d5, 0x000000000000000000000000fe1b2865b74200dceabf9ddee04fdc32c2c8e3fb], data: Bytes([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2]), block_hash: Some(0xa7da1b9509e381c56f9305d86f668fed9476adca16dd02c0c5ff3d36a5fefe3e), block_number: Some(9), transaction_hash: Some(0x0adb0c2be6c040f5dd2aad51aa0b09b4d45a6d6a12b748592ebfb3ddad54b450), transaction_index: Some(0), log_index: Some(0), transaction_log_index: None, log_type: None, removed: None }
[2019-06-26T09:53:01Z DEBUG witnet_ethereum_bridge] Got witnet block: Block { block_header: BlockHeader { version: 0, beacon: CheckpointBeacon { checkpoint: 541235, hash_prev_block: 1693493add74aac1689dcd2485a8c3261e90ba35054fe76fb2fea78e5a382cde }, merkle_roots: BlockMerkleRoots { mint_hash: d1bfb11fc4c8546a01d448b823e2eb57d0eecef6844eadfe968d483c7f16fcbb, vt_hash_merkle_root: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855, dr_hash_merkle_root: eda132c1eb3afac8d6c83c777feaf5173cff502ec6625552af4f2151d331f294, commit_hash_merkle_root: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855, reveal_hash_merkle_root: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855, tally_hash_merkle_root: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855 }, proof: BlockEligibilityClaim { proof: VrfProof { proof: [3, 237, 172, 212, 11, 56, 35, 218, 83, 104, 67, 143, 139, 117, 250, 82, 137, 1, 114, 32, 118, 189, 108, 126, 111, 101, 136, 122, 197, 90, 89, 222, 132, 140, 4, 185, 148, 135, 44, 244, 44, 76, 86, 111, 79, 17, 37, 162, 191, 203, 144, 47, 86, 140, 204, 11, 101, 219, 26, 130, 120, 223, 104, 165, 74, 211, 110, 152, 159, 204, 195, 73, 11, 183, 61, 125, 175, 185, 195, 160, 157], public_key: PublicKey { compressed: 3, bytes: [158, 86, 59, 137, 32, 92, 60, 155, 16, 72, 31, 121, 206, 86, 91, 135, 63, 36, 240, 13, 80, 234, 254, 248, 98, 240, 51, 120, 75, 72, 199, 29] } } } }, block_sig: KeyedSignature { signature: Secp256k1(Secp256k1Signature { der: [48, 68, 2, 32, 78, 210, 117, 141, 15, 193, 113, 189, 61, 141, 110, 0, 34, 246, 140, 59, 240, 237, 102, 66, 193, 237, 13, 8, 92, 28, 51, 127, 52, 137, 247, 152, 2, 32, 64, 131, 31, 163, 191, 25, 161, 130, 179, 184, 115, 180, 167, 43, 169, 20, 128, 164, 136, 231, 86, 249, 10, 7, 125, 167, 19, 66, 125, 177, 129, 224] }), public_key: PublicKey { compressed: 3, bytes: [158, 86, 59, 137, 32, 92, 60, 155, 16, 72, 31, 121, 206, 86, 91, 135, 63, 36, 240, 13, 80, 234, 254, 248, 98, 240, 51, 120, 75, 72, 199, 29] } }, txns: BlockTransactions { mint: MintTransaction { epoch: 541235, output: ValueTransferOutput { pkh: PublicKeyHash { hash: [13, 164, 177, 9, 192, 137, 38, 5, 137, 12, 13, 134, 152, 252, 113, 253, 211, 22, 51, 241] }, value: 50000000000 }, hash: MemoHash { hash: Cell { value: None } } }, value_transfer_txns: [], data_request_txns: [DRTransaction { body: DRTransactionBody { inputs: [Input { output_pointer: 5f0215a96ee91f7f52d294b332318e48cb69bb07692da44cba51a58b60f53c91:0 }], outputs: [ValueTransferOutput { pkh: PublicKeyHash { hash: [67, 103, 66, 172, 250, 40, 99, 246, 83, 194, 251, 57, 95, 157, 209, 84, 63, 243, 158, 31] }, value: 49999998998 }], dr_output: DataRequestOutput { data_request: RADRequest { not_before: 0, retrieve: [RADRetrieve { kind: HttpGet, url: "http://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8", script: [149, 83, 204, 132, 146, 1, 164, 100, 97, 116, 97, 204, 128, 146, 1, 0] }], aggregate: RADAggregate { script: [145, 146, 102, 32] }, consensus: RADConsensus { script: [145, 146, 102, 32] }, deliver: [RADDeliver { kind: HttpGet, url: "https://hooks.zapier.com/hooks/catch/3860543/l2awcd/" }, RADDeliver { kind: HttpGet, url: "https://hooks.zapier.com/hooks/catch/3860543/l1awcw/" }] }, value: 1002, witnesses: 2, backup_witnesses: 1, commit_fee: 0, reveal_fee: 0, tally_fee: 0, time_lock: 0 }, hash: MemoHash { hash: Cell { value: None } } }, signatures: [KeyedSignature { signature: Secp256k1(Secp256k1Signature { der: [48, 69, 2, 33, 0, 154, 50, 107, 238, 181, 155, 53, 216, 185, 97, 245, 67, 94, 215, 206, 24, 74, 133, 7, 44, 14, 1, 64, 82, 176, 171, 235, 18, 208, 103, 195, 174, 2, 32, 113, 127, 12, 248, 29, 17, 182, 79, 193, 54, 69, 152, 161, 166, 131, 79, 173, 76, 223, 46, 235, 145, 157, 96, 132, 150, 254, 20, 34, 183, 25, 121] }), public_key: PublicKey { compressed: 3, bytes: [72, 74, 168, 219, 78, 21, 163, 208, 250, 222, 158, 97, 228, 147, 66, 170, 57, 229, 40, 142, 158, 3, 176, 144, 130, 143, 225, 123, 45, 63, 60, 167] } }] }], commit_txns: [], reveal_txns: [], tally_txns: [] } }
[2019-06-26T09:53:01Z DEBUG witnet_ethereum_bridge] Proof of inclusion for data request eda132c1eb3afac8d6c83c777feaf5173cff502ec6625552af4f2151d331f294:
Data: [10, 206, 1, 18, 76, 18, 56, 104, 116, 116, 112, 58, 47, 47, 113, 114, 110, 103, 46, 97, 110, 117, 46, 101, 100, 117, 46, 97, 117, 47, 65, 80, 73, 47, 106, 115, 111, 110, 73, 46, 112, 104, 112, 63, 108, 101, 110, 103, 116, 104, 61, 49, 38, 116, 121, 112, 101, 61, 117, 105, 110, 116, 56, 26, 16, 149, 83, 204, 132, 146, 1, 164, 100, 97, 116, 97, 204, 128, 146, 1, 0, 26, 6, 10, 4, 145, 146, 102, 32, 34, 6, 10, 4, 145, 146, 102, 32, 42, 54, 18, 52, 104, 116, 116, 112, 115, 58, 47, 47, 104, 111, 111, 107, 115, 46, 122, 97, 112, 105, 101, 114, 46, 99, 111, 109, 47, 104, 111, 111, 107, 115, 47, 99, 97, 116, 99, 104, 47, 51, 56, 54, 48, 53, 52, 51, 47, 108, 50, 97, 119, 99, 100, 47, 42, 54, 18, 52, 104, 116, 116, 112, 115, 58, 47, 47, 104, 111, 111, 107, 115, 46, 122, 97, 112, 105, 101, 114, 46, 99, 111, 109, 47, 104, 111, 111, 107, 115, 47, 99, 97, 116, 99, 104, 47, 51, 56, 54, 48, 53, 52, 51, 47, 108, 49, 97, 119, 99, 119, 47, 16, 234, 7, 24, 2, 32, 1]
TxInclusionProof { index: 0, lemma: [8401056693a26722bec0861042bb4374588eb6fa4893253761869e830a1e301d] }
```
---------------------------------------------------------------------------------------------------------------

### CODE REVIEW

**Comment by lrubiorod** [comment]
# How to run a Witnet-Ethereum bridge node

## Ethereum (local)

The following steps are for local testing with Ganache.
For instructions on how to run in a public Ethereum testnet, see below.

    ganache-cli -b 10 -m scatter clap print else pioneer goat monster clay mystery question deposit measure

This will start an Ethereum client at 127.0.0.1:8545
The `-b 10` flag is important because it sets the block time to 10 seconds.
This allows us to wait for confirmations instead of blindly assuming that the
transactions get accepted.

Copy the first account to witnet-rust/witnet_ethereum_bridge.toml as eth_account
It must be the first account if we want to use the BlockRelay, since currently
the only account that can relay Witnet blocks to Ethereum is the one that
deployed the BlockRelay contract, which by default is the first account.

In order to deploy the contracts, clone the [witnet-ethereum-bridge](https://github.com/witnet/witnet-ethereum-bridge)
Enter the `witnet-ethereum-bridge` directory and run

    truffle deploy --network local

If the command results in an error complaining about "network local does not exist", just add it to `truffle-config.js`, under networks:

    local: {
      host: "127.0.0.1",     // Localhost (default: none)
      port: 8545,            // Standard Ethereum port (default: none)
      network_id: "*",       // Any network (default: none)
    },


Copy BlockRelay contract address to witnet-rust/witnet_ethereum_bridge.toml
Copy WBI contract address to witnet-rust/witnet_ethereum_bridge.toml

If the api of the contracts was modified recently, you have to manually copy
the ABI from

    witnet-ethereum-bridge/build/contracts/BlockRelay.json

(only the contents of the "abi" key of the JSON) to

    witnet-rust/bridges/ethereum/block_relay_abi.json

(the contents of block_relay_abi.json should be a JSON array)

Same for the WBI: copy

    witnet-ethereum-bridge/build/contracts/WitnetBridgeInterface.json

to

    witnet-rust/bridges/ethereum/wbi_abi.json

## Ethereum (Ropsten)

First you have to synchronize with a ethereum node. It is recommended use a light node.

    ./geth --testnet --syncmode "light" --cache=2048 --rpc --rpcport=8545 --rpcapi=eth,web3,net,personal --allow-insecure-unlock

After your node is synced, open a new terminal and connect with the node

    ./geth attach <ethereum_path>/.ethereum/testnet/geth.ipc

Create a new account and unlocked it:

    personal.newAccount(<password>)
    personal.unlockAccount(<account>, <password>, 999999)

Copy this account to witnet-rust/witnet_ethereum_bridge.toml as eth_account. To deploy the contracts it is necessary first to send Eth to this account (using Metamask for example)

In order to deploy the contracts, clone the [witnet-ethereum-bridge](https://github.com/witnet/witnet-ethereum-bridge)
Enter the `witnet-ethereum-bridge` directory and run

    truffle deploy --network ropsten

If the command results in an error complaining about "network local does not exist", just add it to `truffle-config.js`, under networks:

    ropsten: {
      network_id: 3,       // Ropsten's id
      host: "127.0.0.1",   // Localhost (default: none)
      port: 8545,          // Standard Ethereum port (default: none)
      gas: 5500000,        // Ropsten has a lower block limit than mainnet
      confirmations: 2,    // # of confs to wait between deployments. (default: 0)
      timeoutBlocks: 200,  // # of blocks before a deployment times out  (minimum/default: 50)
      skipDryRun: true     // Skip dry run before migrations? (default: false for public nets )
    },


Copy BlockRelay contract address to witnet-rust/witnet_ethereum_bridge.toml
Copy WBI contract address to witnet-rust/witnet_ethereum_bridge.toml

If the api of the contracts was modified recently, you have to manually copy
the ABI from

    witnet-ethereum-bridge/build/contracts/BlockRelay.json

(only the contents of the "abi" key of the JSON) to

    witnet-rust/bridges/ethereum/block_relay_abi.json

(the contents of block_relay_abi.json should be a JSON array)

Same for the WBI: copy

    witnet-ethereum-bridge/build/contracts/WitnetBridgeInterface.json

to

    witnet-rust/bridges/ethereum/wbi_abi.json
## Witnet

Run the node:

    RUST_LOG=witnet=debug cargo run -- -c witnet.toml node run

This will start a JSONRPC client at 127.0.0.1:21338
This node will connect to the Witnet testnet. Wait until the node is SYNCED.
For the correct functionality, the node should have an account with enough
value to post the data requests. The simplest way to achieve that is to mine
a block.

## Bridge

Start the bridge:

    RUST_LOG=witnet_ethereum_bridge=debug cargo run -p witnet-ethereum-bridge

It will try to subscribe to blocks from Witnet and WBI events from Ethereum.

If that's successful, we can post a data request to the WBI:

    cargo run -p witnet-ethereum-bridge --example post_dr

This will send a data request querying the price of bitcoin.

* The bridge will be listening to events from the WBI, so when the transaction
gets accepted into a block, the WBI contract will emit a `PostDataRequest` event.

* The bridge will call `claimDataRequests` in order to claim the data request,
and when that transaction is accepted, the bridge will post the data request to
Witnet.

* When this data request is included in a Witnet block, the bridge will send the
proof of inclusion to the WBI contract. This will emit an `InclusionDataRequest`
event, which indicates to all the bridge nodes that they should start checking
all the new Witnet blocks for a tally which resolves that data request.

* Once a tally has been included in a block, any bridge node can sent the proof
of inclusion. If that proof is valid, the WBI contract will emit a `PostResult`
event indicating that the data request has been resolved.

## Block Relay

A crucial component needed for the correct functionality of the bridge is the
block relay: an Ethereum contract that stores the headers of Witnet blocks.
The current version of the bridge also acts as a block relay, but that can be
disable in the configuration file.

File: bridges/ethereum/examples/post_dr.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+fn eth_event_stream(
+    config: Arc<Config>,
+    eth_state: Arc<EthState>,
+) -> impl Future<Item = (), Error = ()> {
+    let wbi_contract = eth_state.wbi_contract.clone();
+
+    let data_request_output = data_request_example();
+
+    let tally_value = U256::from_dec_str("50000000000000000").unwrap();
```
Comment (mariocao) [inline bridges/ethereum/examples/post_dr.rs]:
```suggestion
    let tally_value = U256::from_dec_str("50_000_000_000_000_000").unwrap();
```

File: bridges/ethereum/examples/post_dr.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    let build_dr_str = include_str!("../../../examples/bitcoin_price.json");
+    let build_dr: serde_json::Value = serde_json::from_str(build_dr_str).unwrap();
+    let mut data_request_output: DataRequestOutput =
+        serde_json::from_value(build_dr["params"]["dro"].clone()).unwrap();
+    data_request_output.data_request.not_before = since_the_epoch.as_secs();
+
+    data_request_output
+}
+
+fn eth_event_stream(
```
Comment (mariocao) [inline bridges/ethereum/examples/post_dr.rs]: I missed small comments describing what the function is doing and some small description of the used values (e.g. gas)

File: bridges/ethereum/examples/post_dr.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    let tally_value = U256::from_dec_str("50000000000000000").unwrap();
+    let data_request_bytes = data_request_output.to_pb_bytes().unwrap();
+
+    wbi_contract
+        .call(
+            "postDataRequest",
+            (data_request_bytes, tally_value),
+            config.eth_account,
+            contract::Options::with(|opt| {
+                opt.value = Some(U256::from_dec_str("250000000000000000").unwrap());
```
Comment (mariocao) [inline bridges/ethereum/examples/post_dr.rs]:
```suggestion
                opt.value = Some(U256::from_dec_str("250_000_000_000_000_000").unwrap());
```

**Comment by mariocao** [comment]
Superseded by #722

**Pull request closed without merging**